File: scripts/fix-test-errors.py
#!/usr/bin/env python3
"""
Fix test file TypeScript errors
"""

import os
import re
from pathlib import Path

PROJECT_ROOT = Path(__file__).parent.parent

def fix_test_file_errors(file_path: Path):
    """Fix common test file errors"""
    try:
        with open(file_path, 'r', encoding='utf-8') as f:
            content = f.read()
    except Exception:
        return False

    original_content = content

    # Fix: Cannot assign to 'NODE_ENV' because it is a read-only property
    # Replace process.env.NODE_ENV = ... with Object.defineProperty
    content = re.sub(
        r'process\.env\.NODE_ENV\s*=\s*([^;]+);',
        lambda m: f'Object.defineProperty(process.env, "NODE_ENV", {{ value: {m.group(1)}, writable: true, configurable: true }});',
        content
    )

    # Fix: Cannot assign to 'url' because it is a read-only property
    content = re.sub(
        r'(\w+)\.url\s*=\s*([^;]+);',
        lambda m: f'Object.defineProperty({m.group(1)}, "url", {{ value: {m.group(2)}, writable: true, configurable: true }});',
        content
    )

    # Fix: Cannot assign to 'method' because it is a read-only property
    content = re.sub(
        r'(\w+)\.method\s*=\s*([^;]+);',
        lambda m: f'Object.defineProperty({m.group(1)}, "method", {{ value: {m.group(2)}, writable: true, configurable: true }});',
        content
    )

    # Fix: Cannot find name 'afterEach'
    if 'afterEach' in content and 'from \'vitest\'' not in content and 'from "vitest"' not in content:
        if 'import' in content:
            # Check if there's already a vitest import
            vitest_import = re.search(r'^import\s+\{([^}]+)\}\s+from\s+[\'"]vitest[\'"];?$', content, re.MULTILINE)
            if vitest_import:
                # Add afterEach to existing import
                imports = vitest_import.group(1)
                if 'afterEach' not in imports:
                    new_imports = imports.rstrip() + ', afterEach'
                    content = content.replace(vitest_import.group(0), f'import {{ {new_imports} }} from \'vitest\';')
            else:
                # Add new import
                import_pattern = re.compile(r'^import .* from [\'"].*[\'"];?$', re.MULTILINE)
                imports = list(import_pattern.finditer(content))
                if imports:
                    last_import = imports[-1]
                    insert_pos = last_import.end()
                    content = content[:insert_pos] + "\nimport { afterEach } from 'vitest';" + content[insert_pos:]
        else:
            # Add at top
            content = "import { afterEach } from 'vitest';\n" + content

    if content != original_content:
        with open(file_path, 'w', encoding='utf-8') as f:
            f.write(content)
        return True
    
    return False

def main():
    """Main function to fix test file errors"""
    
    test_dir = PROJECT_ROOT / '__tests__'
    
    if not test_dir.exists():
        print("No __tests__ directory found")
        return
    
    fixed_files = []
    
    for file_path in test_dir.rglob('*.ts'):
        if file_path.is_file():
            print(f"Processing: {file_path.relative_to(PROJECT_ROOT)}")
            
            fixed = fix_test_file_errors(file_path)
            
            if fixed:
                fixed_files.append(file_path)
                print(f"  ✓ Fixed: {file_path.relative_to(PROJECT_ROOT)}")
    
    print(f"\n✓ Fixed {len(fixed_files)} test files")

if __name__ == '__main__':
    main()

File: scripts/fix-test-errors.js
#!/usr/bin/env node
/**
 * Fix all test file TypeScript errors
 */

const fs = require('fs');
const path = require('path');

// Fix specific test files
const fixes = [
  {
    file: '__tests__/api/menu.test.ts',
    changes: [
      {
        find: 'const request = new Request(',
        replace: 'const request = new NextRequest(',
      },
    ],
  },
  {
    file: '__tests__/api/orders.test.ts',
    changes: [
      {
        find: /mockRequest\.url = 'http:\/\/localhost:3000\/api\/orders\?venueId=venue-1&status=COMPLETED';/g,
        replace: 'Object.defineProperty(mockRequest, \'url\', { value: \'http://localhost:3000/api/orders?venueId=venue-1&status=COMPLETED\', writable: true });',
      },
      {
        find: /mockRequest\.method = 'POST';/g,
        replace: 'Object.defineProperty(mockRequest, \'method\', { value: \'POST\', writable: true });',
      },
    ],
  },
  {
    file: '__tests__/hooks/useMenuItems.test.ts',
    changes: [
      {
        find: /it\('should fetch menu items', async \(\) => \{[\s\S]*?expect\(result\.current\.menuItems\)\.toHaveLength\(2\);/g,
        replace: (match) => {
          return match.replace(/expect\(result\.current\.menuItems\)\.toHaveLength\(2\);/g, 'expect(result.current.menuItems).toHaveLength(2);\n      return true;');
        },
      },
    ],
  },
  {
    file: '__tests__/logger/production-logger.test.ts',
    changes: [
      {
        find: 'afterEach(() => {',
        replace: 'afterEach(() => {',
        addImport: 'import { afterEach } from \'vitest\';',
      },
      {
        find: /process\.env\.NODE_ENV = 'production';/g,
        replace: 'Object.defineProperty(process.env, \'NODE_ENV\', { value: \'production\', writable: true });',
      },
      {
        find: /process\.env\.NODE_ENV = 'development';/g,
        replace: 'Object.defineProperty(process.env, \'NODE_ENV\', { value: \'development\', writable: true });',
      },
    ],
  },
  {
    file: '__tests__/middleware/authorization.test.ts',
    changes: [
      {
        find: /const context = \{[\s\S]*?venue: \{ venue_id: 'venue-123' \},[\s\S]*?user: \{ id: 'user-123' \},[\s\S]*?role: 'owner',[\s\S]*?venueId: 'venue-123'[\s\S]*?\};/g,
        replace: 'const context: AuthorizedContext = {\n      venue: {\n        venue_id: \'venue-123\',\n        owner_id: \'user-123\',\n        name: \'Test Venue\',\n        created_at: new Date().toISOString(),\n        updated_at: new Date().toISOString(),\n      },\n      user: { id: \'user-123\' },\n      role: \'owner\',\n      venueId: \'venue-123\'\n    };',
      },
    ],
  },
];

fixes.forEach(({ file, changes }) => {
  const fullPath = path.join(__dirname, '..', file);
  
  if (!fs.existsSync(fullPath)) {
    console.log(`Skipping ${file} (not found)`);
    return;
  }
  
  let content = fs.readFileSync(fullPath, 'utf8');
  let modified = false;
  
  changes.forEach(({ find, replace, addImport }) => {
    if (typeof find === 'string') {
      if (content.includes(find)) {
        if (typeof replace === 'function') {
          content = content.replace(find, replace);
        } else {
          content = content.replace(find, replace);
        }
        modified = true;
      }
    } else if (find instanceof RegExp) {
      if (content.match(find)) {
        if (typeof replace === 'function') {
          content = content.replace(find, replace);
        } else {
          content = content.replace(find, replace);
        }
        modified = true;
      }
    }
    
    if (addImport && !content.includes(addImport)) {
      content = addImport + '\n' + content;
      modified = true;
    }
  });
  
  if (modified) {
    fs.writeFileSync(fullPath, content, 'utf8');
    console.log(`✅ Fixed ${file}`);
  }
});

console.log('\n✅ All test errors fixed!');

